src/librustc_typeck/check/mod.rs

```diff
@@ -3709,7 +3709,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             self.consider_hint_about_removing_semicolon(blk, expected_ty, err);
                         }
                         if let Some(fn_span) = fn_span {
-                            err.span_label(fn_span, "this function's body doesn't return");
+                            err.span_label(
+                                fn_span,
+                                "this function implicitly returns `()` as its body has no tail \
+                                 expression nor a `return` statement",
```

Comment:
```suggestion
                                 nor `return`-expression",
```

Reply:
Is the current

```rust
                                "this function implicitly returns `()` as its body has no tail \
                                 or `return` expression",
```

reasonable or would you like me to change it to the requested wording? I have an irrational aversion to compound-words.

Reply:
Seems reasonable to me as-is. The main advantage of writing *"tail- nor `return`-expression"* is that it clarifies that it refers to *tail expression nor `return` expression* without the repetition.